### PR TITLE
feat(fstring): treat fstring as string literal

### PR DESCRIPTION
**Pull request by kurikomoe**
I found the output of `ic(f"var: {var}")` is quite annoying, so I made some changes.

> Add includeFStringExpressions options to ic to control whether print
>   fstring expressions.
> Change README.md for includeFStringExpressions

f-string: [PEP-498](https://www.python.org/dev/peps/pep-0498/)

For python below 3.6, the f-string expression will not pass the syntax check, so I didn't add any python version checks in function `isFString(s)`.

Also, I'm not quite sure about how to implement unit tests in icecream repo. Please add tests if necessary.

### CODE REVIEW

**Comment by alexmojaki** [comment]
Can you explain when this is useful? Why not just write `ic(var)` or `print(f"var: {var}")`?

**Comment by kurikomoe** [comment]
1. print(f"val: {val}")
I would like to use `ic` to output debug information and tend not to mixed using the ic and print.
After all, `IceCream — Never use print() to debug again`


2. `ic(val)` just cannot contains enough information:

Some situations:

```pycon
# To show the val1 vs val2 in comparsion rather than put them in two lines. and  reduce output length.
>>> ic(f"shoud be the same - val1: {val1},  val2: {val2}")  
ic| should be the same - val1: Good, val2: Bad
```

```pycon
# To get extra info about the variable
>>> ic(f"Resnet Input Shape: {x.shape}")
Resnet Input Shape: [2, 100, 2000]
# In this case, just ouput x.shape: [2, 100, 2000] will make debug much harder cause there may be tons of x.shape
```

**Comment by alexmojaki** [comment]
> ```
> # To show the val1 vs val2 in comparsion rather than put them in two lines. and  reduce output length.
> >>> ic(f"shoud be the same - val1: {val1},  val2: {val2}")  
> ic| should be the same - val1: Good, val2: Bad
> ```

You can write `ic(val1, val2)`

> ```python
> def forward(self, x):
>     x = self.fc1(x)
>     ic(f"after fc1: {x.shape}")
>     x = self.fc2(x)
>     ic(f"after fc2: {x.shape}")
>     x = self.fc3(x)
>     ic(f"after fc2: {x.shape}")
>     x = torch.tanh(x)
> ```

You can write `ic("fc1", x.shape)` if you want to 'label' your ic calls.

Another possibility is `ic.configureOutput(includeContext=True)` so that it shows the line number. It's not great, but at least the three calls to `ic(x.shape)` can be distinguished by line number.

If you write many `ic()` calls and/or you want to monitor changes to a single variable/expression then I'd recommend https://github.com/alexmojaki/snoop instead. Then you could just write:

```python
import snoop

@snoop
def forward(self, x):
    x = self.fc1(x)
    x = self.fc2(x)
    x = self.fc3(x)
    x = torch.tanh(x)
```

It'll even show `x.shape` automatically. And if you want to mix that with something like `ic` you can use `snoop.pp` which is equivalent.

For PyTorch specifically, there's also https://github.com/zasdfgbnm/TorchSnooper which by default uses PySnooper (also has the slogan "Never use print for debugging again") but can be configured to use snoop.

**Comment by salabim** [comment]
May I suggest adding a values_only/vo parameter, like
    ic(f"var={var}", values_only=True)
or
    ic(f"var={var}", vo=True)

**Comment by kurikomoe** [comment]
> May I suggest adding a values_only/vo parameter, like
> ic(f"var={var}", values_only=True)
> or
> ic(f"var={var}", vo=True)

I agree with this, but for myself, there would be plenty of  `values_only = True` for fstrings

 

> > ```
> > # To show the val1 vs val2 in comparsion rather than put them in two lines. and  reduce output length.
> > >>> ic(f"shoud be the same - val1: {val1},  val2: {val2}")  
> > ic| should be the same - val1: Good, val2: Bad
> > ```
> 
> 
> You can write `ic(val1, val2)`
> 
> 
> > ```python
> > def forward(self, x):
> >     x = self.fc1(x)
> >     ic(f"after fc1: {x.shape}")
> >     x = self.fc2(x)
> >     ic(f"after fc2: {x.shape}")
> >     x = self.fc3(x)
> >     ic(f"after fc2: {x.shape}")
> >     x = torch.tanh(x)
> > ```
> 
> 
> You can write `ic("fc1", x.shape)` if you want to 'label' your ic calls.
> 
> 
> Another possibility is `ic.configureOutput(includeContext=True)` so that it shows the line number. It's not great, but at least the three calls to `ic(x.shape)` can be distinguished by line number.
> 
> 
> If you write many `ic()` calls and/or you want to monitor changes to a single variable/expression then I'd recommend https://github.com/alexmojaki/snoop instead. Then you could just write:
> 
> 
> ```python
> import snoop
> 
> @snoop
> def forward(self, x):
>     x = self.fc1(x)
>     x = self.fc2(x)
>     x = self.fc3(x)
>     x = torch.tanh(x)
> ```
> 
> 
>     
>   
> 
> It'll even show `x.shape` automatically. And if you want to mix that with something like `ic` you can use `snoop.pp` which is equivalent.
> 
> 
> For PyTorch specifically, there's also https://github.com/zasdfgbnm/TorchSnooper which by default uses PySnooper (also has the slogan "Never use print for debugging again") but can be configured to use snoop.

Thanks for PySnoopers and other tools, I will check them later.
But I still think that giving the ability to use f-string to annotation the debug output is essential, and output something like:
```pycon
>>> ic(f"val={val}")
ic | f"val={val}": val=123 
``` 
is quite stupid. I don't think when using f-string, people will care much about the original string expression.

**Comment by gruns** [comment]
@kurikomoe i agree with alex: the goal of icecream is to avoid typing `ic(f"val={val}")` to begin with, replaced with just `ic(val)`

for desired output like

```python
>>> ic(f"shoud be the same - val1: {val1},  val2: {val2}")  
ic| should be the same - val1: Good, val2: Bad
```

is this

```python
>>> ic('should be the same', val1, val2)
```

awkward or counter-intuitive?

i dont disagree with your desire to avoid printing f-strings as both variable name + value, but i want to understand your desires more before merging this. because for me, and my personal uses of `ic()`, wanting to type `ic(f"shoud be the same - val1: {val1},  val2: {val2}")` instead of `ic(val1, val2)` or `ic(val1, val2, val1 == val2)` feels awkward or stretched

**Comment by salabim** [comment]
What about 
```
    x = 12.3
    ic(f"{x:0.3e}")
```
if someone wants to print a number in exponential format?

**Comment by kurikomoe** [comment]
TL;DR. f-string should behave like strings in `ic` because the purpose of f-string is to generate a string.


Then, let me break up my thoughts about f-strings and `ic`

1. I would like to replace `print` with `ic` to get more control of the debug output, and try not to mix using them.
2. When debugging, I indeed need some simple string annotations to help me understand the execution flow.
```pycon
ic('reached here')
....
ic('reached there')
```
3. Based on the 1&2, when I need some more detailed annotations, I use f-string, and then `ic` fails me.
```pycon
ic(f' reached here {cnt}')
....
ic(f'incoming message: {msg}, len: {len(msg)}')
```

You see, I use f-string because the original string is not strong enough to get better annotation or other things.

For simply printing some variables, I agree that`ic(vals)` is quite enough. but the problem is not about variables, but about strings.

In short, when output some strings, especially for `"Tody is {}".format(...)` or  `f"Some Debug {Messages}"`, `ic` is not user friendly for now.

If not for the compatible issues, I even want to suggest making `includeFStringExpression=False` as default

**Comment by alexmojaki** [comment]
Earlier I suggested `ic("fc1", x.shape)` instead of `ic(f"fc1: {x.shape}")`. @gruns suggested `ic('should be the same', val1, val2)` instead of `ic(f"shoud be the same - val1: {val1},  val2: {val2}")`. In both cases I don't see any response to these suggestions. Again, instead of:

```python
ic(f' reached here {cnt}')
....
ic(f'incoming message: {msg}, len: {len(msg)}')
```

why not:

```python
ic(' reached here', cnt)
....
ic('incoming', msg, len(msg))
```

The closest thing to a response that I see is:

> I still think that giving the ability to use f-string to annotation the debug output is essential

and

> I use f-string because the original string is not strong enough to get better annotation or other things.

but it's not clear why you claim this and why a mixture of string literals and other arguments is not a good method of annotation.

I think @gruns is right that if that method doesn't seem good to you it's worth thinking about what that might imply about icecream in general. I'd also add that adding more options to `configureOutput` comes at a cost of making the library more complicated and difficult for users to learn. The options already have some issues and there's other plans to change them, so additional proposals need to prove their value.

**Comment by alexmojaki** [comment]
> I would like to replace print with ic to get more control of the debug output, and try not to mix using them.

This is also an important point. Not wanting to mix them is fine, but as I said we're thinking about changes to `configureOutput` like #31 so it's good to gather data about how `ic` is used and where the output goes. Do you have any special output configuration? Or is it just nice to have everything on one stream (stderr) with the consistent `ic |` prefix?

**Comment by kurikomoe** [comment]
Never mind, this is just a suggestion. I will use my own fork to suspend f-string output.

**Comment by flight505** [comment]
I also have several cases where I just want ic to output a nice f-string only, perhaps with a time stamp. 
Now I have a logger, and ic imported in cases where I need both, silly we can't have the option. 

An example, I just want to know what the current state is: 
logging.info(
        "Update state {key} with value", {st.session_state[key + '_widget']}) 

- too bad the commit hasn't been merged.
